### PR TITLE
EthMiner crate

### PR DESCRIPTION
**Pull request by tomusdrw**
Same as #632 but to master
Please merge #640 & #650 & #651 first.

### CODE REVIEW

**Comment by tomusdrw** [comment]
We should probably wait with this until #641 #609 #587 are merged.
I will resolve all conflicts in a single batch.

**Comment by gavofyork** [comment]
presumably this kind of reverses/rewrites #641, no?

File: ethcore/src/client/client.rs [hunk 1]
```diff
@@ -357,52 +348,60 @@ impl<V> Client<V> where V: Verifier {
...
-	/// Set the extra_data that we will seal blocks with.
-	pub fn set_extra_data(&self, extra_data: Bytes) {
-		*self.extra_data.write().unwrap() = extra_data;
+	fn try_seal(&self, block: ClosedBlock, seal: Vec<Bytes>) -> Result<SealedBlock, ClosedBlock> {
+		block.try_seal(self.engine.deref().deref(), seal)
 	}
 
-	/// New chain head event. Restart mining operation.
-	pub fn prepare_sealing(&self) {
+	fn prepare_sealing(&self, author: Address, extra_data: Bytes, transactions: Vec<SignedTransaction>) -> Option<ClosedBlock> {
```
Comment (gavofyork) [inline ethcore/src/client/client.rs]: will be nice to move this function into the ethminer crate eventually - that can happen in civility though.

**Comment by gavofyork** [comment]
looks like this has some of the other recent changes in. can we get a proper diff to review?

**Comment by tomusdrw** [comment]
It contains #661 and #640 which has been reviewed and merged to this branch.

This PR only introduces ethminer crate and includes transactions to mined block. I can split it into couple of PRs if you prefer (but it basicly requires reverting commits and reverting the reverts in separate prs)

**Comment by gavofyork** [comment]
ahh i see. that rather makes it hard to see what changes have been reviewed and what are new :-/

File: miner/src/miner.rs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+				.block(BlockId::Hash(*hash))
+				// Client should send message after commit to db and inserting to chain.
+				.expect("Expected in-chain blocks.");
+			let block = BlockView::new(&block);
+			block.transactions()
+		}
+
+		{
+			let good = good.par_iter().map(|h| fetch_transactions(chain, h));
+			let bad = bad.par_iter().map(|h| fetch_transactions(chain, h));
```
Comment (gavofyork) [inline miner/src/miner.rs]: is this really "bad" (i.e. invalid)? or just retracted? if retracted, then rest of logic is fine. if invalid, then rest of logic is broken (we don't want to be adding transactions from invalid blocks).

**Comment by gavofyork** [comment]
1000+ lines changed over 32 files. this is too big, and more importantly, diffuse. the fact that some arbitrary set of those changes have already been reviewed doesn't help a bit, since we can't tell which parts.

this includes various refactorings, some small, some big. i've reviewed it as best as i can and the style and basic premise is good, but there is no way i would sign off on the logic or there being no unintended consequences.

would have been much nicer if we had incremental alterations where each step can be logic-checked.

**Pull request closed without merging**